### ISSUE TITLE
`disallowed_methods` can't detect higher-order calls

### ISSUE THREAD

**Issue by michaelsproul** [rust-lang/rust-clippy#8849]
### Summary

The `disallowed_methods` lint is unable to detect calls to higher-order functions. Consider this `clippy.toml` which bans `String::with_capacity` and the following code which Clippy fails to flag:

```toml
disallowed-methods = [
	"std::string::String::with_capacity",
]
```

```rust
#![deny(clippy::disallowed_methods)]

fn main() {
    let caps = vec![1, 2, 3];
    let strings = caps
        .into_iter()
        .map(std::string::String::with_capacity)
        .collect::<Vec<_>>();
    println!("{:?}", strings);
}
```

The problem is that Clippy doesn't recognise that passing `String::with_capacity` to `Iterator::map` will result in it being called. This seems to me like a _very_ difficult problem to solve, because it requires whole program analysis -- we need to analyse every call to every dependency, including `std`. As far as I know, the information about which functions are called by a specific instantiation of a generic function is not available in HIR, and lints can't run over a crate and all of its dependent crates anyway.

There are a wide variety of ways to invoke a function in a higher-order or indirect way (see the Reproducer section for another example).

I wonder if we would be better doing this analysis in a different tool with access to later compiler passes. Should we do it in MIR? In the LLVM bytecode? In the machine code binary (this seems insane)?

### Lint Name

disallowed_methods

### Reproducer

```rust
#![deny(clippy::disallowed_methods)]

pub struct Indirect<T, U> {
    f: Box<dyn FnOnce(T) -> U>,
    g: Box<dyn FnOnce(T) -> U>,
}

impl<T, U> Indirect<T, U> {
    fn go(self, t: T, selector: bool) -> U {
        if selector {
            (self.f)(t)
        } else {
            (self.g)(t)
        }
    }
}

fn main() {
    let indirect = Indirect {
        f: Box::new(std::string::String::with_capacity),
        g: Box::new(|_| String::new()),
    };
    let s = indirect.go(5, true);
    println!("{}", s.capacity());
}
```

Outcome: Clippy raises no warnings.

Expected: `indirect.go(5, true)` should be disallowed because it could result in (and does result in) a call to the disallowed function `String::with_capacity`.

### Version

```text
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-unknown-linux-gnu
release: 1.61.0
LLVM version: 14.0.0
```

**Comment by michaelsproul** [comment]
Maybe it's not as complicated as I initially thought. We could ignore the harder problem of determining whether a crate _calls_ a function and just determine whether it _refers_ to a disallowed function. That would catch both my examples, and I think many other variations 🤔

**Comment by Alexendoo** [comment]
Ah timing, I had just had that thought also and it seemed to work fine so I submitted the PR. I can't think of any reason you'd want to refer to a disallowed method but not call it, so hopefully that means no FPs
